### PR TITLE
Fix coding rate send chunk rounding error

### LINKED ISSUE #85: Strange results in wifi 802.11
Hello,

I know I was the one who submitted the code in first place, and as far as I know, I implemented it according to the IEEE 802.11 standard.
According to the IEEE 802.11 section 17.3.5.6 the data is coded with a convolutional encoded of rate 1/2. Higher rates are achieved by then puncturing. In the receiver side, there is depuncturing by adding a dummy "zero" on the place of the omitted zeros. 

This works good in some MCS (0,1,2,3,4) where the parameters have coding of 1/2 and 3/4 and modulation BPSK, QPSK and 16-QAM.
Graph with the BER vs SNR bellow (message size was 600bits, 2000 runs each SNR):
image

When higher MCS are tried (5, 6,7, 8, 9) the behaviour becomes strange as can be seen in a similar plot for those MCS bellow (same conditions of experiment):
image

- The first and more visible problem is with MCS 6, after a particular SNR (about 24) the BER stops decreasing at the same rate as others MCS (it still decreases but only slightly, maybe rounding differences).
- Second the MCS 7 achieves better BER than MCS 5 at most SNRs, notice both are 64-QAM, the only difference are the punturing performed. Similar behaviour happens with MCS 9 in relation with MCS 8.

Does someone have any idea why this simulation results are happening? 
Some error in code or some error in my interpretation of the results?

If it helps debug, I notice some warnings in my console:

> /commpy/commpy/modulation.py:131: RuntimeWarning: divide by zero encountered in double_scalars
>   demod_bits[i * self.num_bits_symbol + self.num_bits_symbol - 1 - bit_index] = log(llr_num / llr_den)

This is maybe related with the problem with MCS 6 at SNRs over 25, I can easily replicate this issue.

Also, I can add code if it helps. :-)

### PR DESCRIPTION
**Pull request by eSoares**
Fixes #85

### CODE REVIEW

**Comment by BastienTr** [comment]
Thanks for you PR @eSoares. I didn't run your test but this make a lot of sense.

However, I'm not sure of why you are using Fraction rather than regular float? Is it just to compute the denominator that is used to determine `divider`?

**Comment by eSoares** [comment]
> However, I'm not sure of why you are using Fraction rather than regular float? Is it just to compute the denominator that is used to determine `divider`?

Fraction helps reducing the usage of floats and keeps numbers in integers. 
I was noticing that every once in a while, a float like 0.75 or 0.5 when converting to fraction was getting huge numbers as numerators and denominators due the floating point errors. Letting the parameter be a Fraction completely avoids any issues that can happen with that.

**Comment by BastienTr** [comment]
> I was noticing that every once in a while, a float like 0.75 or 0.5 when converting to fraction was getting huge numbers as numerators and denominators due the floating point errors. Letting the parameter be a Fraction completely avoids any issues that can happen with that.

I understand. However, I can't found out why you need to convert float (e.g. 0.5) to fraction. Why is a very small floating-point error an issue? Is it due to the integer division?

**Comment by eSoares** [comment]
> > I was noticing that every once in a while, a float like 0.75 or 0.5 when converting to fraction was getting huge numbers as numerators and denominators due the floating point errors. Letting the parameter be a Fraction completely avoids any issues that can happen with that.
> 
> I understand. However, I can't found out why you need to convert float (e.g. 0.5) to fraction. Why is a very small floating-point error an issue? Is it due to the integer division?

Is due to the divider calculation in https://github.com/veeresht/CommPy/pull/86/commits/c5919c2cf94ab541c7bbe3a18ef892657c9f2dd2#diff-346d73f899a78915b2bb8779e25ccc49R212 .

**Comment by BastienTr** [comment]
Ok and the rounding error here leads to the error that are observed in #85. If so, congratulation for spotting a so tricky bug! I'll merge this PR as soon as you confirm this.

Thanks :+1:

File: commpy/links.py [hunk 1]
```diff
@@ -200,18 +203,21 @@ def link_performance_full_metrics(self, SNRs, tx_max, err_min, send_chunk=None,
         BEs = np.zeros((len(SNRs), tx_max), dtype=int)  # Bit errors per tx
         CEs = np.zeros((len(SNRs), tx_max), dtype=int)  # Chunk Errors per tx
         NCs = np.zeros((len(SNRs), tx_max), dtype=int)  # Number of Chunks per tx
-        # Set chunk size and round it to be a multiple of num_bits_symbol*nb_tx to avoid padding
+        # Set chunk size and round it to be a multiple of num_bits_symbol* nb_tx to avoid padding taking in to account the coding rate
         if send_chunk is None:
             send_chunk = err_min
-        divider = self.num_bits_symbol * self.channel.nb_tx
```
Comment (BastienTr) [inline commpy/links.py:206]: Ok. Then, it seems to me that changing this line with the following should fixe #85 without requiering Fraction. Are there, rounding errors in this line?

`divider = self.num_bits_symbol * self.channel.nb_tx / code_rate`

Comment (eSoares) [reply commpy/links.py:206]: That line (and/or the next one) is indeed the centre of the question, and where it needs "the fix".
But that change is not the same that I'm suggesting. That finds the send chunk that when applied the coding rate is divisible by num_bits_symbol, but does not ensure in the receiver, after demodulation, the number of bits before decoding are divisible by the code_rate.

Taking the example of my previous comment, the issue would still exist.
The result send_chunk  with that code would be `(600//(6/(3/4)))*6/(3/4) = 600`under the code that I have suggested, the result would be `594 = (600//(6*3)) * (6*3)`, 594 after coding 3/4 will be 792 which is divisible by 6.

**Comment by BastienTr** [comment]
Ok I think I get it. Thanks for providing the solution to this issue ! :smiley:

**Pull request merged**